renderdoc/driver/gl/gl_hooks_vrapi.cpp

```diff
@@ -0,0 +1,312 @@
+          libvrapi_symHandle, "vrapi_GetTextureSwapChainHandle");
+    if(vrapi_GetSystemPropertyInt_real == NULL)
+      vrapi_GetSystemPropertyInt_real =
+          (PFN_vrapi_GetSystemPropertyInt)dlsym(libvrapi_symHandle, "vrapi_GetSystemPropertyInt");
+
+    return vrapi_SubmitFrame_real != NULL;
+  }
+
+} vrapi_hooks;
+
+//-----------------------------------------------------------------------------------------------------------------
+GLenum GetInternalFormat(ovrTextureFormat ovr_format)
+{
+  GLenum internalFormat = eGL_RGBA8;
+
+  GLenum conversion_table[] = {
+      eGL_RGBA8,                // VRAPI_TEXTURE_FORMAT_NONE,
+      eGL_RGB565,               // VRAPI_TEXTURE_FORMAT_565,
+      eGL_RGB5_A1,              // VRAPI_TEXTURE_FORMAT_5551,
+      eGL_RGBA4,                // VRAPI_TEXTURE_FORMAT_4444,
+      eGL_RGBA8,                // VRAPI_TEXTURE_FORMAT_8888,
+      eGL_SRGB8_ALPHA8,         // VRAPI_TEXTURE_FORMAT_8888_sRGB,
+      eGL_RGBA16F,              // VRAPI_TEXTURE_FORMAT_RGBA16F,
+      eGL_DEPTH_COMPONENT16,    // VRAPI_TEXTURE_FORMAT_DEPTH_16,
+      eGL_DEPTH_COMPONENT24,    // VRAPI_TEXTURE_FORMAT_DEPTH_24,
+      eGL_DEPTH24_STENCIL8      // VRAPI_TEXTURE_FORMAT_DEPTH_24_STENCIL_8,
+  };
+
+  RDCASSERT(ovr_format < ARRAY_COUNT(conversion_table));
+
+  if(ovr_format < ARRAY_COUNT(conversion_table))
+  {
+    internalFormat = conversion_table[ovr_format];
+  }
+
+  return internalFormat;
+}
+
+GLenum GetTextureType(ovrTextureType ovr_tex_type)
+{
+  GLenum textureType = eGL_TEXTURE_2D;
+
+  GLenum conversion_table[] = {
+      eGL_TEXTURE_2D,          // VRAPI_TEXTURE_TYPE_2D,
+      eGL_TEXTURE_2D,          // VRAPI_TEXTURE_TYPE_2D_EXTERNAL,
+      eGL_TEXTURE_2D_ARRAY,    // VRAPI_TEXTURE_TYPE_2D_ARRAY,
+      eGL_TEXTURE_CUBE_MAP     // VRAPI_TEXTURE_TYPE_CUBE
+  };
+
+  RDCASSERT(ovr_tex_type < ARRAY_COUNT(conversion_table));
+
+  if(ovr_tex_type < ARRAY_COUNT(conversion_table))
+  {
+    textureType = conversion_table[ovr_tex_type];
+  }
+
+  return textureType;
+}
+
+//-----------------------------------------------------------------------------------------------------------------
+extern "C" {
+
+__attribute__((visibility("default"))) ovrMobile *vrapi_EnterVrMode(const ovrModeParms *parms)
+{
+  if(vrapi_hooks.vrapi_EnterVrMode_real == NULL)
+  {
+    vrapi_hooks.SetupHooks();
+  }
+
+  ovrMobile *ovr = vrapi_hooks.vrapi_EnterVrMode_real(parms);
+
+  if(m_GLDriver)
+  {
+    void *ctx = m_GLDriver->GetCtx();
+    void *wndHandle = ovr;
+
+    RenderDoc::Inst().AddFrameCapturer(ctx, wndHandle, m_GLDriver);
+    RenderDoc::Inst().SetActiveWindow(ctx, wndHandle);
+
+#if ENABLED(RDOC_ANDROID)
+    if(parms->Flags & VRAPI_MODE_FLAG_NATIVE_WINDOW)
+    {
+      ANativeWindow *window = (ANativeWindow *)parms->WindowSurface;
+
+      int32_t width = ANativeWindow_getWidth(window);
+      int32_t height = ANativeWindow_getHeight(window);
+
+      m_GLDriver->WindowSize(ovr, width, height);
+    }
+#endif
```

Comment:
I don't think it is a good idea to have this code here. The window size should be set to the size of the actual backbuffer associated with the context, and it's not clear if this `WindowSurface` must be the same as the `EGLSurface` used to create the current context. If there's a mismatch in size then that could lead to errors since we'll try to both read from an invalid size on capture to save the thumbnail screenshot, but also on replay we won't create the right dimension of texture to substitute as the backbuffer dimensions.

You mentioned earlier that without this code the thumbnail was 8x8, but I don't think this is the right place to fix that. If the thumbnail was coming out as 8x8 then that suggests that the real backbuffer is indeed that size, since the default is 32x32 before it gets provided any dimensions from in `eglSwapBuffers`. If it's not the right size then you'd need to trace through the handling of the EGL surface itself - since that's how the size should be controlled to make sure everything is consistent.